File: frontend/src/api/stores/get-stats.ts
import { api, CatalogEndpoints } from '@/api/utils';
import { queryOptions, useQuery, type UseQueryOptions, type UseQueryResult } from '@tanstack/react-query';

export type StatsType = 'revenue' | 'visitors' | 'customers' | 'all';

export type ChartDataPoint = {
  date: string;
  revenue: number;
  sales: number;
  customers: number;
  visitors: number;
};

export type ProductDataPoint = {
  productSlug: string;
  name: string;
  sales: number;
  revenue: number;
};

export type StatsTotals = {
  revenue: number;
  sales: number;
  customers: number;
  visitors: number;
  averageRevenue: number;
  averageSales: number;
  conversionRate: string;
};

export type StatsResponse = {
  chartData: ChartDataPoint[];
  productData: ProductDataPoint[];
  totals: StatsTotals;
};

export type GetStatsParams = {
  type?: StatsType;
  startsAt?: string; // ISO date string
  endsAt?: string; // ISO date string
};

async function getStats(params?: GetStatsParams): Promise<StatsResponse> {
  const queryParams: Record<string, string | undefined> = {
    type: params?.type,
    startsAt: params?.startsAt,
    endsAt: params?.endsAt,
  };

  const res = await api.get<StatsResponse>(CatalogEndpoints.Statistics.GetStats, {
    params: queryParams,
  });
  return res;
}

export const StatsKeys = {
  all: () => ['stats'] as const,
  filtered: (params?: GetStatsParams) => ['stats', params] as const,
};

export function useGetStats(
  params?: GetStatsParams,
  overrides?: Partial<UseQueryOptions<StatsResponse, Error>>,
): UseQueryResult<StatsResponse, Error> {
  return useQuery<StatsResponse, Error>(
    queryOptions({
      queryFn: () => getStats(params),
      queryKey: StatsKeys.filtered(params),
      ...overrides,
    }),
  );
}

File: backend/Booking/src/Modules/Booking.Modules.Catalog/Features/Statistics/GetStats/GetStats.cs
using Booking.Common;
using Booking.Common.Authentication;
using Booking.Common.Endpoints;
using Booking.Common.Messaging;
using Booking.Common.Results;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Routing;

namespace Booking.Modules.Catalog.Features.Statistics.GetStats;

internal sealed class GetStats : IEndpoint
{
    public void MapEndpoint(IEndpointRouteBuilder app)
    {
        app.MapGet(CatalogEndpoints.Statistics.GetStats, async (
                [FromQuery] string? type,
                [FromQuery] DateTime? startsAt,
                [FromQuery] DateTime? endsAt,
                UserContext userContext,
                IQueryHandler<GetStatsQuery, StatsResponse> handler,
                CancellationToken cancellationToken) =>
            {
                int userId = userContext.UserId;
                var query = new GetStatsQuery(userId, type, startsAt, endsAt);

                var result = await handler.Handle(query, cancellationToken);

                return result.Match(
                    Results.Ok,
                    CustomResults.Problem);
            })
            .RequireAuthorization()
            .WithTags(Tags.Statistics);
    }
}

public record GetStatsQuery(
    int UserId,
    string? Type,
    DateTime? StartsAt,
    DateTime? EndsAt) : IQuery<StatsResponse>;

public record StatsResponse(
    List<ChartDataPoint> ChartData,
    List<ProductDataPoint> ProductData,
    StatsTotals Totals);

public record ChartDataPoint(
    string Date,
    decimal Revenue,
    int Sales,
    int Customers,
    int Visitors);

public record ProductDataPoint(
    string ProductSlug,
    string Name,
    int Sales,
    decimal Revenue);

public record StatsTotals(
    decimal Revenue,
    int Sales,
    int Customers,
    int Visitors,
    decimal AverageRevenue,
    decimal AverageSales,
    string ConversionRate);
